tt_metal/impl/dispatch/device_command.cpp

```diff
@@ -12,104 +12,126 @@
+    CQPrefetchCmd *relay_wait = (CQPrefetchCmd *)((char *)this->cmd_region + this->cmd_write_offsetB);
     if (!zero_init_disable) DeviceCommand::zero(relay_wait);
 
-    relay_wait.base.cmd_id = CQ_PREFETCH_CMD_RELAY_INLINE;
-    relay_wait.relay_inline.length = sizeof(CQDispatchCmd);
-    relay_wait.relay_inline.stride = CQ_PREFETCH_CMD_BARE_MIN_SIZE;
+    relay_wait->base.cmd_id = CQ_PREFETCH_CMD_RELAY_INLINE;
+    relay_wait->relay_inline.length = sizeof(CQDispatchCmd);
+    relay_wait->relay_inline.stride = CQ_PREFETCH_CMD_BARE_MIN_SIZE;
 
-    this->write_to_cmd_sequence(&relay_wait, sizeof(CQPrefetchCmd));
+    this->cmd_write_offsetB += sizeof(CQPrefetchCmd);
 
-    CQDispatchCmd wait_cmd;
+    CQDispatchCmd *wait_cmd = (CQDispatchCmd *)((char *)this->cmd_region + this->cmd_write_offsetB);
     if (!zero_init_disable) DeviceCommand::zero(wait_cmd);
 
-    wait_cmd.base.cmd_id = CQ_DISPATCH_CMD_WAIT;
-    wait_cmd.wait.barrier = barrier;
-    wait_cmd.wait.notify_prefetch = false;
-    wait_cmd.wait.addr = address;
-    wait_cmd.wait.count = count;
-    wait_cmd.wait.clear_count = clear_count;
+    wait_cmd->base.cmd_id = CQ_DISPATCH_CMD_WAIT;
+    wait_cmd->wait.barrier = barrier;
+    wait_cmd->wait.notify_prefetch = false;
+    wait_cmd->wait.addr = address;
+    wait_cmd->wait.count = count;
+    wait_cmd->wait.clear_count = clear_count;
 
-    this->write_to_cmd_sequence(&wait_cmd, sizeof(CQDispatchCmd));
+    this->cmd_write_offsetB += sizeof(CQDispatchCmd);
 }
 
 void DeviceCommand::add_dispatch_wait_with_prefetch_stall(uint8_t barrier, uint32_t address, uint32_t count, uint8_t clear_count) {
     this->add_dispatch_wait(barrier, address, count, clear_count);
 
-    uint32_t dispatch_wait_idx = this->cmd_write_idx - (sizeof(CQDispatchCmd) / sizeof(uint32_t));
-    CQDispatchCmd *wait_cmd = (CQDispatchCmd*)(this->cmd_sequence.data() + dispatch_wait_idx);
+    uint32_t dispatch_wait_offsetB = this->cmd_write_offsetB - sizeof(CQDispatchCmd);
+    CQDispatchCmd *wait_cmd = (CQDispatchCmd*)((char *)this->cmd_region + dispatch_wait_offsetB);
     wait_cmd->wait.notify_prefetch = true;
 
-    CQPrefetchCmd stall_cmd;
+    uint32_t increment_sizeB = align(sizeof(CQPrefetchCmd), PCIE_ALIGNMENT);
+    this->validate_cmd_write(increment_sizeB);
+
+    CQPrefetchCmd *stall_cmd = (CQPrefetchCmd *)((char *)this->cmd_region + this->cmd_write_offsetB);
     if (!zero_init_disable) DeviceCommand::zero(stall_cmd);
 
-    stall_cmd.base.cmd_id = CQ_PREFETCH_CMD_STALL;
+    stall_cmd->base.cmd_id = CQ_PREFETCH_CMD_STALL;
 
-    this->write_to_cmd_sequence(&stall_cmd, sizeof(CQPrefetchCmd), PCIE_ALIGNMENT);
+    this->cmd_write_offsetB += increment_sizeB;
 }
 
 void DeviceCommand::add_prefetch_relay_inline(bool flush, uint32_t lengthB) {
-    CQPrefetchCmd relay_write;
+    this->validate_cmd_write(sizeof(CQPrefetchCmd));
+    CQPrefetchCmd *relay_write = (CQPrefetchCmd *)((char *)this->cmd_region + this->cmd_write_offsetB);
```

Comment:
minor: maybe an inline templated access routine to do the cast?  
  relay_write = this->reserve_space<CQPrefetchCmd *>(sizeof(CQPrefetchCmd));
this would do the validate on space and return the pointer?  could also do the zero init...

Reply:
yeah good call